adoc/chapters/programming_interface.adoc

```diff
@@ -17640,40 +17635,500 @@ member function template.
+
+#ifdef SYCL_SIMPLE_SWIZZLES
+__swizzled_vec__</*unspecified*/> XYZW_SWIZZLE() const
+__swizzled_vec__</*unspecified*/> RGBA_SWIZZLE() const
+#endif
+
+__swizzled_vec__</*unspecified*/> lo() const
+__swizzled_vec__</*unspecified*/> hi() const
+__swizzled_vec__</*unspecified*/> odd() const
+__swizzled_vec__</*unspecified*/> even() const
+----
+!====
+_Constraints:_ These functions have the same constraints as the equivalent
+member functions of the [code]#vec# class.
+
+_Returns:_ A new [code]#+__swizzled_vec__+# view of the underlying [code]#vec#
+object, where the view represents the combination of two swizzle operations.
+The first is the swizzle operation represented by the "this" view.
+The second is the swizzle operation defined by the member function.
+The indices used by the second swizzle are the indices produced by the first
+swizzle.
+For example, if the second swizzle references the first element, this means the
+element of the underlying [code]#vec# that corresponds to the first element
+produced by the first swizzle.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+template <access::address_space AddressSpace, access::decorated IsDecorated>
+void load(size_t offset, multi_ptr<const DataT, AddressSpace, IsDecorated> ptr)
+----
+!====
+_Effects:_ Loads values from memory into elements of the underlying [code]#vec#
+object.
+A total of [code]#NumElements# values are loaded from memory, starting at the
+the address [code]#ptr + offset*sizeof(DataT)#.
+The first value from memory is written to the element in [code]#vec# that
+corresponds to the first element of the swizzle operation.
+The second value from memory is written to the element in [code]#vec# that
+corresponds to the second element of the swizzle operation, etc.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+DataT& operator[](int index)
+const DataT& operator[](int index) const
+----
+!====
+_Returns:_ A reference to the element of the underlying [code]#vec# object that
+corresponds to the position [code]#index# of the swizzle operation.
+
+a@
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+template</*unspecified*/>
+__swizzled_vec__& operator=(const __swizzled_vec__</*unspecified*/>& rhs)
+----
+!====
+_Constraints:_ Available only when the element data type of [code]#rhs# is the
+same as [code]#DataT# and when the number of elements in the [code]#rhs# view
+is equal to [code]#NumElements#.
+
+_Effects:_ Assigns elements from the [code]#rhs# [code]#+__swizzled_vec__+#
+view to elements of "this" [code]#+__swizzled_vec__+# view.
+The value from the [code]#rhs# corresponding to the first element of the
+[code]#rhs# swizzle operation is assigned to the element of the underlying
+[code]#vec# object that corresponds to the first element of the "this" swizzle
+operation, etc.
+
+The order in which the elements are assigned is unspecified.
+Therefore, if the "this" view has any repeated elements in the swizzle, those
```

Comment:
I think this wording may need to be stronger than this, the original wording did not permit swizzles on the lhs of an expression to have repeated indexes.

Reply:
So the intent of the original wording was to restrict the interface such that a swizzle with repeated elements could not be used on the lhs of an expression. So an expected implementation of this would be that the relevant operators for a swizzle with repeated elements are not available, resulting in a compile-time error for the user. I agree, this should have been more clearly specified in the wording of the operators and member functions.

The rationale for this approach was that the `vec` class was based on [Clang ext_vector_type ](https://clang.llvm.org/docs/LanguageExtensions.html#vectors-and-extended-vectors) and for this using repeated elements on the lhs of an expression results in a compile-time error (https://godbolt.org/z/dGz84s6Ge). I think we could leave the behavior unspecified, but I think it would be difficult for this to be supported consistently, one implementation could specify a priority so the result could be reasoned about and avoid UB, but other implementations may throw a compile-time error. I would prefer to not provide the interface, and avoid the user doing this by mistake.

Reply:
OK, that makes sense.  I think we could add this as a constraint to the function specification.  It seems like this could be implemented as a compiler error using normal SFINAE techniques, and that would work for any compiler.

Reply:
Yeah I think this would be a good way to do it.

Reply:
Done in 9585110c44a52a9fb2f81b40652f56c4b05ecac8.